data_transfer/lib/byteflies.py

```diff
@@ -0,0 +1,227 @@
+import json
+import time
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, List, Optional
+
+import requests
+
+from data_transfer.config import config
+from data_transfer.utils import DeviceType, read_csv_from_cache, uid_to_hash
+
+
+@dataclass
+class BytefliesFileDownload:
+    """Use as CLI arguments for ByteFlies' library."""
+
+    directory: Path = config.storage_vol
+    ftype: str = "csv"
+
+
+# Define location and filetype to download
+args = BytefliesFileDownload()
+
+
+def get_token(creds: dict) -> str:
+    """
+    Generates a AWS Cognito IdToken for API access
+    # NOTE: generally expires in 60 minutes
+    # TODO: establish token refreshing depending on \
+        frequency/length of DAG
+    """
+    res = requests.post(
+        f"{config.byteflies_aws_auth_url}",
+        headers={
+            "X-Amz-Target": "AWSCognitoIdentityProviderService.InitiateAuth",
+            "Content-Type": "application/x-amz-json-1.1",
+        },
+        json={
+            "ClientId": f"{creds['client_id']}",
+            "AuthFlow": "USER_PASSWORD_AUTH",
+            "AuthParameters": {
+                "USERNAME": f"{creds['username']}",
+                "PASSWORD": f"{creds['password']}",
+            },
+        },
+    )
+    #  TODO: catch/log exception
+    res.raise_for_status()
+    resp = res.json()
+    return str(resp["AuthenticationResult"]["IdToken"])
+
+
+def get_session(token: str) -> requests.Session:
+    """
+    Builds a requests session object with the required header
+    """
+    session = requests.Session()
+    session.headers.update({"Authorization": f"{token}"})
+    return session
+
+
+def get_list(session: requests.Session, from_date: str, to_date: str) -> List[dict]:
+    """
+    GET a list of records (metadata) across study sites, or 'groups' in ByteFlies API
+    """
+
+    @dataclass
+    class __metadata_copy:
+        """Generator for deepcopying a BTF json object"""
+
+        jsondump: str
+
+        def get(self, signal_id: str, algorithm_id: str = "") -> Any:
+            copy = json.loads(self.jsondump)
+            copy["IDEAFAST"] = {
+                "hash": uid_to_hash(
+                    f"{copy['id']}/{signal_id}/{algorithm_id}", DeviceType.BTF
+                ),
+                "signal_id": signal_id,
+                "algorithm_id": algorithm_id,
+            }
+            return copy
+
+    groups = __get_groups(session)
```

Comment:
might be worth renaming `groups` here and elsewhere to `study_site`. Although they're groups on byteflie's platform, it's easier to think of them as study sites?